ProcessLib/ThermalTwoPhaseFlowWithPP/ThermalTwoPhaseFlowWithPPLocalAssembler-impl.h

```diff
@@ -0,0 +1,473 @@
+    auto Kgpc = local_K.template block<nonwet_pressure_size, cap_pressure_size>(
+        nonwet_pressure_matrix_index, cap_pressure_matrix_index);
+    auto Kgt = local_K.template block<nonwet_pressure_size, temperature_size>(
+        nonwet_pressure_matrix_index, temperature_matrix_index);
+
+    auto Klp = local_K.template block<cap_pressure_size, nonwet_pressure_size>(
+        cap_pressure_matrix_index, nonwet_pressure_matrix_index);
+    auto Klpc = local_K.template block<cap_pressure_size, cap_pressure_size>(
+        cap_pressure_matrix_index, cap_pressure_matrix_index);
+    auto Klt = local_K.template block<cap_pressure_size, temperature_size>(
+        cap_pressure_matrix_index, temperature_matrix_index);
+
+    auto Kep = local_K.template block<temperature_size, nonwet_pressure_size>(
+        temperature_matrix_index, nonwet_pressure_matrix_index);
+    auto Kepc = local_K.template block<temperature_size, cap_pressure_size>(
+        temperature_matrix_index, cap_pressure_matrix_index);
+    auto Ket = local_K.template block<temperature_size, temperature_size>(
+        temperature_matrix_index, temperature_matrix_index);
+
+    auto Bg = local_b.template segment<nonwet_pressure_size>(
+        nonwet_pressure_matrix_index);
+    auto Bl =
+        local_b.template segment<cap_pressure_size>(cap_pressure_matrix_index);
+    auto Be =
+        local_b.template segment<temperature_size>(temperature_matrix_index);
+
+    unsigned const n_integration_points =
+        _integration_method.getNumberOfPoints();
+
+    SpatialPosition pos;
+    pos.setElementID(_element.getID());
+    const int material_id =
+        _process_data.material->getTwoPhaseMaterialModel()->getMaterialID(
+            pos.getElementID().get());
+
+    auto const num_nodes = ShapeFunction::NPOINTS;
+    auto pg_nodal_values =
+        Eigen::Map<const Eigen::VectorXd>(&local_x[0], num_nodes);
+    auto pc_nodal_values =
+        Eigen::Map<const Eigen::VectorXd>(&local_x[num_nodes], num_nodes);
+
+    const Eigen::MatrixXd& perm =
+        _process_data.material->getTwoPhaseMaterialModel()->getPermeability(
+            material_id, t, pos, _element.getDimension());
+    assert(perm.rows() == _element.getDimension() || perm.rows() == 1);
+    GlobalDimMatrixType permeability = GlobalDimMatrixType::Zero(
+        _element.getDimension(), _element.getDimension());
+    if (perm.rows() == _element.getDimension())
+        permeability = perm;
+    else if (perm.rows() == 1)
+        permeability.diagonal().setConstant(perm(0, 0));
+
+    for (unsigned ip = 0; ip < n_integration_points; ip++)
+    {
+        double pg_int_pt = 0.;
+        double pc_int_pt = 0.;
+        double T_int_pt = 0.0;  // Temperature
+        NumLib::shapeFunctionInterpolate(local_x, _ip_data[ip].N, pg_int_pt,
+                                         pc_int_pt, T_int_pt);
+
+        double const rho_water =
+            _process_data.material->getTwoPhaseMaterialModel()
```

Comment:
the material model access happens often in the code, it would be better to create a reference to that model introducing a new variable (maybe faster too).